src/main/proto/stats.proto

```diff
@@ -0,0 +1,319 @@
+  google.protobuf.Timestamp execute_start = 9;
+  google.protobuf.Timestamp execute_end = 10;
+  google.protobuf.Timestamp download_outputs_start = 11;
+  google.protobuf.Timestamp download_outputs_end = 12;
+  // TODO(olaola): add more when local fallback is implemented.
+}
+
+// Properties of an executed action.
+message ActionMetadata {
+  // Number of overall action inputs (input files and other blobs).
+  int32 num_inputs = 1;
+
+  // Total bytes of all the inputs.
+  int64 total_input_bytes = 2;
+
+  // Number of actual produced action outputs (not listed outputs).
+  int32 num_outputs = 3;
+
+  // Total bytes of actual produced outputs.
+  int64 total_output_bytes = 4;
+
+  // Number of inputs that needed to be uploaded to CAS.
+  int32 cas_cache_misses = 5;
+}
+
+// Properties of a file.
+message File {
+  // Path to the file relative to the execution root.
+  string path = 1;
+
+  // Digest of the file's contents.
+  build.bazel.remote.execution.v2.Digest digest = 2;
+}
+
+// Full execution data that can be used to debug cache misses.
+message ExecutionData {
+  // All the action input files, including traversed input directories.
+  repeated File input_files = 1;
+
+  // All the local files that needed to be uploaded to remote CAS.
+  repeated string input_file_cas_misses = 5;
+
+  // Actual produced output files.
+  repeated File output_files = 2;
+
+  // Command proto digest (command proto is one of the input blobs).
+  build.bazel.remote.execution.v2.Digest command_digest = 3;
+
+  // Action proto digest (action proto is one of the input blobs).
+  build.bazel.remote.execution.v2.Digest action_digest = 4;
+}
+
+// The full record of a single command execution.
+message RunRecord {
+  // Execution stages, in topological order.
+  enum Stage {
+    UNKNOWN = 0;
+    QUEUED = 1;
+    COMPUTING_INPUT_TREE = 2;
+    CHECKING_ACTION_CACHE = 3;
+    UPLOADING_INPUTS = 4;
+    EXECUTING = 5;
+    DOWNLOADING_OUTPUTS = 6;
+    LOCAL_FALLBACK_EXECUTING = 7;
+    LOCAL_FALLBACK_UPLOADING_OUTPUTS = 8;
+    FINISHED = 9;
+  }
+  // The current execution stage.
+  Stage stage = 1;
+
+  // Parameters of the command to execute.
+  RunCommandParameters command_parameters = 2;
+
+  // Result, if available.
+  RunResult result = 3;
+
+  // Properties of the action.
+  ActionMetadata action_metadata = 4;
+
+  // Timestamps of the proxy execution.
+  LocalTimestamps local_timestamps = 5;
+
+  // If action was run remotely, these are the timestamps returned by the remote
+  // server.
+  build.bazel.remote.execution.v2.ExecutedActionMetadata remote_metadata = 6;
+
+  // Saved failed remote result, before local re-execution.
+  RunResult result_before_local_fallback = 7;
+
+  // Full execution data, if available (not saved by default).
+  ExecutionData execution_data = 8;
+}
+
+// Used in Stats rpc: query the proxy server about the stats of its current state,
+message StatsRequest {
+  // If set, returns the full list of saved RunRecords.
```

Comment:
should this be an enum?

Reply:
I don't think so, they are independent and do different things. The only reason  said one should be set is that otherwise you'd get an empty response. But let me rename these so that they are clearer, and reuse the new Labels message instead.

Reply:
So you can set both?

Reply:
Oh, yes, I do set both when there is only one proxy instance.